flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,317 @@
+                        inputList -> {
+                            DataStream inoutData = inputList.get(0);
+                            return inoutData.map(
+                                    new PredictLabelFunction(
+                                            broadcastKey, getK(), getFeaturesCol()),
+                                    outputTypeInfo);
+                        });
+
+        return new Table[] {tEnv.fromDataStream(output)};
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return this.params;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+        ReadWriteUtils.saveModelData(
+                KnnModelData.getModelDataStream(modelDataTable),
+                path,
+                new KnnModelData.ModelDataEncoder(),
+                0);
+    }
+
+    /**
+     * Loads model data from path.
+     *
+     * @param env Stream execution environment.
+     * @param path Model path.
+     * @return Knn model.
+     */
+    public static KnnModel load(StreamExecutionEnvironment env, String path) throws IOException {
+        KnnModel model = ReadWriteUtils.loadStageParam(path);
+        Table modelDataTable = ReadWriteUtils.loadModelData(env, path, new ModelDataDecoder(), 0);
+        return model.setModelData(modelDataTable);
+    }
+
+    /** This operator loads model data and predicts result. */
+    private static class PredictLabelFunction extends RichMapFunction<Row, Row> {
+        private boolean firstElement = true;
+        private final String featureCol;
+        private transient KnnModelData knnModelData;
+        private final int topN;
+        private final String broadcastKey;
+        private transient Comparator<? super Tuple2<Double, Double>> comparator;
+
+        public PredictLabelFunction(String broadcastKey, int k, String featureCol) {
+            this.topN = k;
+            this.broadcastKey = broadcastKey;
+            this.featureCol = featureCol;
+        }
+
+        @Override
+        public Row map(Row row) {
+            if (firstElement) {
+                comparator = Comparator.comparingDouble(o -> -o.f0);
+                loadModel(getRuntimeContext().getBroadcastVariable(broadcastKey));
+                firstElement = false;
+            }
+            DenseVector vector = (DenseVector) row.getField(featureCol);
+            Tuple2<List<Double>, List<Double>> t2 = findNeighbor(vector, topN);
+            return Row.join(row, Row.of(getResult(t2)));
+        }
+
+        /**
+         * Finds the nearest topN neighbors from whole nodes.
+         *
+         * @param input Input vector.
+         * @param topN Top N.
+         * @return Neighbors.
+         */
+        private Tuple2<List<Double>, List<Double>> findNeighbor(DenseVector input, Integer topN) {
+            PriorityQueue<Tuple2<Double, Double>> priorityQueue = new PriorityQueue<>(comparator);
```

Comment:
Since we are OK to instantiate `priorityQueue` every time the `map` is invoked, could we also instantiate `comparator` here as well, for better code simplicity and consistency?

Reply:
OK